src/validators/AllowedSessionsValidator.sol

```diff
@@ -0,0 +1,152 @@
+  mapping(bytes32 sessionActionsHash => bool active) public areSessionActionsAllowed;
+
+  /// @notice Emitted when session actions are allowed or disallowed.
+  /// @param sessionActionsHash The hash of the session actions.
+  /// @param allowed Boolean indicating if the session actions are allowed.
+  event SessionActionsAllowed(bytes32 indexed sessionActionsHash, bool indexed allowed);
+
+  constructor() {
+    _grantRole(SESSION_REGISTRY_MANAGER_ROLE, msg.sender);
+    _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+  }
+
+  /// @notice Set whether a session actions hash is allowed or not.
+  /// @param sessionActionsHash The hash of the session actions.
+  /// @param allowed Boolean indicating if the session actions are allowed.
+  /// @dev Session actions represent the set of operations, such as fee limits, call policies, and transfer policies,
+  /// that define the behavior and constraints of a session.
+  function setSessionActionsAllowed(
+    bytes32 sessionActionsHash,
+    bool allowed
+  ) external virtual onlyRole(SESSION_REGISTRY_MANAGER_ROLE) {
+    if (areSessionActionsAllowed[sessionActionsHash] != allowed) {
+      areSessionActionsAllowed[sessionActionsHash] = allowed;
+      emit SessionActionsAllowed(sessionActionsHash, allowed);
+    }
+  }
+
+  /// @notice Get the hash of session actions from a session specification.
+  /// @param sessionSpec The session specification.
+  /// @return The hash of the session actions.
+  /// @dev The session actions hash is derived from the session's fee limits, call policies, and transfer policies.
+  function getSessionActionsHash(SessionLib.SessionSpec memory sessionSpec) public view virtual returns (bytes32) {
+    bytes32 feeLimitAndTransferPoliciesHash = keccak256(abi.encode(sessionSpec.feeLimit, sessionSpec.transferPolicies));
+    bytes32 callPoliciesHash;
+    {
+      uint256 len = sessionSpec.callPolicies.length;
+      if (len > 0) {
+        // Each CallSpec has: address (32), bytes4 (32), uint256 (32), UsageLimit (3*32) = 128 bytes for first 4 fields
+        // But UsageLimit is a struct of 3 uint256, so 96 bytes for UsageLimit
+        // So, total: 32 + 32 + 32 + 96 = 192 bytes per CallSpec (first 4 fields)
+        uint256 fieldsSize = 32 + 32 + 32 + 96;
+        bytes memory buf = new bytes(len * fieldsSize);
+        for (uint256 i = 0; i < len; ++i) {
+          SessionLib.CallSpec memory c = sessionSpec.callPolicies[i];
+          uint256 offset = i * fieldsSize;
+          assembly {
+            mstore(add(buf, add(32, offset)), mload(c)) // target
+            mstore(add(buf, add(64, offset)), mload(add(c, 32))) // selector
+            mstore(add(buf, add(96, offset)), mload(add(c, 64))) // maxValuePerUse
+            // valueLimit is a struct of 3 uint256
+            let valueLimit := mload(add(c, 96))
+            mstore(add(buf, add(128, offset)), mload(valueLimit)) // limitType
```

Comment:
Here we are effectively storing `mload(mload(add(c, 96))` which does not look right. 

```suggestion
            let valueLimit := add(c, 96)
            mstore(add(buf, add(128, offset)), mload(valueLimit)) // limitType
```

As assembly does go with a higher risk of bugs, any reason to employ it here? Are there any significant gas savings? We could do something like

```solidity
bytes[] memory callPolicies = new bytes[](len);
for (uint256 i = 0; i < len; ++i) {
    SessionLib.CallSpec memory policy = sessionSpec.callPolicies[i];
    callPolicies[i] = abi.encode(policy.target, policy.selector, policy.maxValuePerUse, policy.valueLimit);
}
```

And hash that?

Reply:
Ah, just for the optimizations sake, they could be significant if a lot of call policies would be introduced, but, I think you're right, it's truly an overengineering thing. I'll fix that right away =)

Reply:
And yes, that was a typo, my bad =) Thanks!

> Here we are effectively storing `mload(mload(add(c, 96)))` which does not look right.